Wb 1396 merge objects retain properties

MartinImgSpc opened the pull request:
Basically in the past all objects were merged to new ProBuilderMesh. Right now we re-use the active mesh as the first target to re-use.

This address WB-1372 and WB-1373 as well.

thomas-tu commented on Runtime/MeshOperations/CombineMeshes.cs:
```diff
@@ -114,6 +114,196 @@ public static List<ProBuilderMesh> Combine(IEnumerable<ProBuilderMesh> meshes)
+        /// A list of merged meshes. In most cases this will be a single mesh corresponding to meshTarget. However it can be multiple in cases
+        /// where the resulting vertex count exceeds the maximum allowable value.
+        /// </returns>
+        public static List<ProBuilderMesh> Combine(ProBuilderMesh meshTarget, IEnumerable<ProBuilderMesh> meshes)
```
I would suggest to re-use `List<ProBuilderMesh> Combine(IEnumerable<ProBuilderMesh> meshes)` logic then inject the result into a given target. To me, the logic stays the same, only where we store the first mesh data changes (and we still generate new ProBuilderMeshes if we exceed the  vertex count limit).

MartinImgSpc replied:
It is not totally the same as we need to convert the other mesh instances into the first target space and we do not want to alter the target internal data as we are processing those, which minimizes the manipulation. Furthermore, if we exceed the vertex count we are using the old method, except in the case if there is a single mesh remaining to merge which would fail the call(I am unsure in this case what would be the benefit of reprocessing it to generate a new mesh?)

Am I missing your point?
Thank you!

thomas-tu replied:
I see what you mean but I feel the code has some redundant parts and could probably be divided better or refactored. For example, I would have maybe taken the opportunity to offer a new function to Combine a mesh as much as it can with a list of meshes within the limit of vertex count like if you were creating a texture atlas. Then use `List<ProBuilderMesh> Combine (IEnumerable<ProBuilderMesh> meshes)` to combine the remaining meshes like you did at the end.

On a side note, I also don't see the benefit with having the first loop if the sole idea is to inject the meshTarget data  into `vertices`, `sharedVertices`, `sharedTextures` and `materialMap`. Maybe I'm missing something. Otherwise, I would go with `FirstOrDefault()` to check if the object exists in the passed collection.

Maybe what I'm proposing is too overkill. @karl- , thoughts?

MartinImgSpc replied:
@thomas-tu  I definitely agree with your comment on the first loop. Will unloop.

karl- replied:
I agree that we should not have two separate functions for combining meshes.

My thoughts on this are we should make the behavior of keeping the first source mesh gameobject and all it's components the default, removing the current path.

thomas-tu commented on Runtime/MeshOperations/CombineMeshes.cs:
```diff
@@ -114,6 +114,196 @@ public static List<ProBuilderMesh> Combine(IEnumerable<ProBuilderMesh> meshes)
+            meshTarget.Refresh();
+
+            var returnedMesh = new List<ProBuilderMesh>() { meshTarget };
+            if (remainderMeshContributors.Count > 1)
```
I wonder if this is necessary as I would expect the original Combine() function to handle both cases count > 1 and count == 1.

MartinImgSpc replied:
Actually it does not.
if (!meshes.Any() || meshes.Count() < 2)
                return null;

thomas-tu replied:
Oh.

@karl- , it doesn't return a single mesh if we pass a single mesh?

karl- replied:
No, it should return `null` if the function does not perform any action.

thomas-tu commented on Runtime/MeshOperations/CombineMeshes.cs:
```diff
@@ -136,56 +136,16 @@ public static List<ProBuilderMesh> Combine(ProBuilderMesh meshTarget, IEnumerabl
-                    {
-                        materialMap.Add(mat);
-                    }
+            if (!meshes.Contains(meshTarget))
```
@karl-, @MartinImgSpc:
- Am I right to see this function as this following?
> Here a list of meshes to combine, please use this one as the host/starting point/mesh of reference/target

If so, it would make sense to enforce this condition. Also, I would change its signature to help better understand what it does and to keep a consistent API:
```C#
List<ProBuilderMesh> Combine(IEnumerable<ProBuilderMesh> meshes, ProBuilderMesh meshTarget)
```

- @karl- , is it fine to add a new API function?

karl- replied:
Yes, ok to add a new func signature with the `ProBuilderMesh` as an overload.

MartinImgSpc commented:
@karl- @thomas-tu  I have obsoleted the old Combine method and refactored the code to minimize duplication. Changed the order of the arguments of Combine and updated documentation.

gabrielw-us commented:
Issue:
1. Create pb cube a and pb cube b
2. Deselect all
3. click select a
4. CTRL-click select b (now both are selected)
5. SHIFT-click a (both are still selected, a becomes active item)
6. Use PB Merge

Expected: Cube A is used as the "use my info" cube
Result: Handle disappears from both cubes (odd?), must re-click on them to get the handle back, and Cube B has actually been used as "use my info"

MartinImgSpc commented:
> Issue:
> 
> 1. Create pb cube a and pb cube b
> 2. Deselect all
> 3. click select a
> 4. CTRL-click select b (now both are selected)
> 5. SHIFT-click a (both are still selected, a becomes active item)
> 6. Use PB Merge
> 
> Expected: Cube A is used as the "use my info" cube
> Result: Handle disappears from both cubes (odd?), must re-click on them to get the handle back, and Cube B has actually been used as "use my info"

This is an issue with the selection not with this code change.  I am getting the current mesh from MeshSelection.activeMesh which gives me b in the case above. I will investigate why, I suspect we have the handle disappear because it is not returning a.

gabrielw-us commented:
Issue:
1. Create 2 pb cubes and 1 Unity cube
2. Click-select 1 pb cube
3. CTRL click the other pb cube, so both are selected
4. CTRL click the unity cube, now all three cubes are selected
5. Click "Merge" in PB toolbar

Expected: Only the PB items are merged
Result: No merge happens, 2 errors logged:

```
Undo object may not be null.
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

```
ArgumentNullException: Value cannot be null.
Parameter name: meshTarget
UnityEngine.ProBuilder.MeshOperations.CombineMeshes.Combine (System.Collections.Generic.IEnumerable`1[T] meshes, UnityEngine.ProBuilder.ProBuilderMesh meshTarget) (at C:/_git/com.unity.probuilder/Runtime/MeshOperations/CombineMeshes.cs:44)
UnityEditor.ProBuilder.Actions.MergeObjects.DoAction () (at C:/_git/com.unity.probuilder/Editor/MenuActions/Object/MergeObjects.cs:48)
UnityEditor.ProBuilder.MenuAction.DoButton (System.Boolean isHorizontal, System.Boolean showOptions, UnityEngine.Rect& optionsRect, UnityEngine.GUILayoutOption[] layoutOptions) (at C:/_git/com.unity.probuilder/Editor/EditorCore/MenuAction.cs:337)
UnityEditor.ProBuilder.EditorToolbar.OnGUI () (at C:/_git/com.unity.probuilder/Editor/EditorCore/EditorToolbar.cs:368)
UnityEditor.ProBuilder.ProBuilderEditor.OnGUI () (at C:/_git/com.unity.probuilder/Editor/EditorCore/ProBuilderEditor.cs:485)
System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <ad04dee02e7e4a85a1299c7ee81c79f6>:0)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <ad04dee02e7e4a85a1299c7ee81c79f6>:0)
System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) (at <ad04dee02e7e4a85a1299c7ee81c79f6>:0)
UnityEditor.HostView.Invoke (System.String methodName, System.Object obj) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
UnityEditor.HostView.Invoke (System.String methodName) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
UnityEditor.HostView.InvokeOnGUI (UnityEngine.Rect onGUIPosition, UnityEngine.Rect viewRect) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
UnityEditor.DockArea.DrawView (UnityEngine.Rect viewRect, UnityEngine.Rect dockAreaRect) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
UnityEditor.DockArea.OldOnGUI () (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
UnityEngine.UIElements.IMGUIContainer.DoOnGUI (UnityEngine.Event evt, UnityEngine.Matrix4x4 parentTransform, UnityEngine.Rect clippingRect, System.Boolean isComputingLayout, UnityEngine.Rect layoutSize, System.Boolean eventIsPropagatedFromNonFocusableVisualElement) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.IMGUIContainer.HandleIMGUIEvent (UnityEngine.Event e, UnityEngine.Matrix4x4 worldTransform, UnityEngine.Rect clippingRect, System.Boolean eventIsPropagatedFromNonFocusableVisualElement) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.IMGUIContainer.HandleIMGUIEvent (UnityEngine.Event e, System.Boolean eventIsPropagatedFromNonFocusableVisualElement) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.IMGUIContainer.SendEventToIMGUI (UnityEngine.UIElements.EventBase evt) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.IMGUIContainer.HandleEvent (UnityEngine.UIElements.EventBase evt) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.CallbackEventHandler.HandleEventAtTargetPhase (UnityEngine.UIElements.EventBase evt) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.MouseCaptureDispatchingStrategy.DispatchEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.EventDispatcher.ProcessEventQueue () (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.EventDispatcher.OpenGate () (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.EventDispatcherGate.Dispose () (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.EventDispatcher.Dispatch (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel, UnityEngine.UIElements.DispatchMode dispatchMode) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.BaseVisualElementPanel.SendEvent (UnityEngine.UIElements.EventBase e, UnityEngine.UIElements.DispatchMode dispatchMode) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.UIElementsUtility.DoDispatch (UnityEngine.UIElements.BaseVisualElementPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.UIElements.UIElementsUtility.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr) (at <3dd62d3208914b14a9c3250b5037e105>:0)
UnityEngine.GUIUtility.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr) (at <47533d5f70c041debc5c751474c7409d>:0)
```

MartinImgSpc commented:
> Issue:
> 
> 1. Create 2 pb cubes and 1 Unity cube
> 2. Click-select 1 pb cube
> 3. CTRL click the other pb cube, so both are selected
> 4. CTRL click the unity cube, now all three cubes are selected
> 5. Click "Merge" in PB toolbar
> 
> Expected: Only the PB items are merged
> Result: No merge happens, 2 errors logged:
> 
> ```
> Undo object may not be null.
> UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
> ```
> 
> ```
> ArgumentNullException: Value cannot be null.
> Parameter name: meshTarget
> UnityEngine.ProBuilder.MeshOperations.CombineMeshes.Combine (System.Collections.Generic.IEnumerable`1[T] meshes, UnityEngine.ProBuilder.ProBuilderMesh meshTarget) (at C:/_git/com.unity.probuilder/Runtime/MeshOperations/CombineMeshes.cs:44)
> UnityEditor.ProBuilder.Actions.MergeObjects.DoAction () (at C:/_git/com.unity.probuilder/Editor/MenuActions/Object/MergeObjects.cs:48)
> UnityEditor.ProBuilder.MenuAction.DoButton (System.Boolean isHorizontal, System.Boolean showOptions, UnityEngine.Rect& optionsRect, UnityEngine.GUILayoutOption[] layoutOptions) (at C:/_git/com.unity.probuilder/Editor/EditorCore/MenuAction.cs:337)
> UnityEditor.ProBuilder.EditorToolbar.OnGUI () (at C:/_git/com.unity.probuilder/Editor/EditorCore/EditorToolbar.cs:368)
> UnityEditor.ProBuilder.ProBuilderEditor.OnGUI () (at C:/_git/com.unity.probuilder/Editor/EditorCore/ProBuilderEditor.cs:485)
> System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <ad04dee02e7e4a85a1299c7ee81c79f6>:0)
> Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
> System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <ad04dee02e7e4a85a1299c7ee81c79f6>:0)
> System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) (at <ad04dee02e7e4a85a1299c7ee81c79f6>:0)
> UnityEditor.HostView.Invoke (System.String methodName, System.Object obj) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
> UnityEditor.HostView.Invoke (System.String methodName) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
> UnityEditor.HostView.InvokeOnGUI (UnityEngine.Rect onGUIPosition, UnityEngine.Rect viewRect) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
> UnityEditor.DockArea.DrawView (UnityEngine.Rect viewRect, UnityEngine.Rect dockAreaRect) (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
> UnityEditor.DockArea.OldOnGUI () (at <2c7108252b3f412abf2ec8c94fb7c234>:0)
> UnityEngine.UIElements.IMGUIContainer.DoOnGUI (UnityEngine.Event evt, UnityEngine.Matrix4x4 parentTransform, UnityEngine.Rect clippingRect, System.Boolean isComputingLayout, UnityEngine.Rect layoutSize, System.Boolean eventIsPropagatedFromNonFocusableVisualElement) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.IMGUIContainer.HandleIMGUIEvent (UnityEngine.Event e, UnityEngine.Matrix4x4 worldTransform, UnityEngine.Rect clippingRect, System.Boolean eventIsPropagatedFromNonFocusableVisualElement) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.IMGUIContainer.HandleIMGUIEvent (UnityEngine.Event e, System.Boolean eventIsPropagatedFromNonFocusableVisualElement) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.IMGUIContainer.SendEventToIMGUI (UnityEngine.UIElements.EventBase evt) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.IMGUIContainer.HandleEvent (UnityEngine.UIElements.EventBase evt) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.CallbackEventHandler.HandleEventAtTargetPhase (UnityEngine.UIElements.EventBase evt) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.MouseCaptureDispatchingStrategy.DispatchEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.EventDispatcher.ProcessEventQueue () (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.EventDispatcher.OpenGate () (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.EventDispatcherGate.Dispose () (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.EventDispatcher.Dispatch (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.IPanel panel, UnityEngine.UIElements.DispatchMode dispatchMode) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.BaseVisualElementPanel.SendEvent (UnityEngine.UIElements.EventBase e, UnityEngine.UIElements.DispatchMode dispatchMode) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.UIElementsUtility.DoDispatch (UnityEngine.UIElements.BaseVisualElementPanel panel) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.UIElements.UIElementsUtility.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr) (at <3dd62d3208914b14a9c3250b5037e105>:0)
> UnityEngine.GUIUtility.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr) (at <47533d5f70c041debc5c751474c7409d>:0)
> ```

Thanks! Pushed a fix as discussed if the activeMesh is not a ProBuilder object MergeObjects will not be enabled.

thomas-tu commented on Runtime/MeshOperations/CombineMeshes.cs:
```diff
@@ -18,7 +18,105 @@ public static class CombineMeshes
         /// A list of merged meshes. In most cases this will be a single mesh. However it can be multiple in cases
         /// where the resulting vertex count exceeds the maximum allowable value.
         /// </returns>
+        [Obsolete("Combine(IEnumerable<ProBuilderMesh> meshes) is deprecated. Plase use Combine(IEnumerable<ProBuilderMesh> meshes, ProBuilderMesh meshTarget).")]
```
Why not keep both Combine(*)?

MartinImgSpc replied:
@thomas-tu  this was done based on suggestion from @karl-  https://github.com/Unity-Technologies/com.unity.probuilder/pull/190#discussion_r313939477

thomas-tu replied:
Oh okey, I missed it, thanks!
